flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveLookupTableSource.java

```diff
@@ -0,0 +1,285 @@
+ * it does not have the lookup ability but can be a temporal table just like other stream sources.
+ * When it acts as bounded table, it has the lookup ability.
+ *
+ * <p>A common user case is use hive table as dimension table and always lookup the latest partition data, in this
+ * case, hive table source is a continuous read source but currently we implements it by LookupFunction. Because
+ * currently TableSource can not tell the downstream when the latest partition has been read finished. This is a
+ * temporarily workaround and will re-implement in the future.
+ */
+public class HiveLookupTableSource extends HiveTableSource implements LookupTableSource {
+
+	private static final Duration DEFAULT_LOOKUP_MONITOR_INTERVAL = Duration.ofHours(1L);
+	private final Configuration configuration;
+	private Duration hiveTableReloadInterval;
+
+	public HiveLookupTableSource(
+			JobConf jobConf,
+			ReadableConfig flinkConf,
+			ObjectPath tablePath,
+			CatalogTable catalogTable) {
+		super(jobConf, flinkConf, tablePath, catalogTable);
+		this.configuration = new Configuration();
+		catalogTable.getOptions().forEach(configuration::setString);
+		validateLookupConfigurations();
+	}
+
+	@Override
+	public LookupRuntimeProvider getLookupRuntimeProvider(LookupContext context) {
+		return TableFunctionProvider.of(getLookupFunction(context.getKeys()));
+	}
+
+	@VisibleForTesting
+	TableFunction<RowData> getLookupFunction(int[][] keys) {
+		int[] keyIndices = new int[keys.length];
+		int i = 0;
+		for (int[] key : keys) {
+			if (key.length > 1) {
+				throw new UnsupportedOperationException("Hive lookup can not support nested key now.");
+			}
+			keyIndices[i] = key[0];
+			i++;
+		}
+		return getLookupFunction(keyIndices);
+	}
+
+	private void validateLookupConfigurations() {
+		String partitionInclude = configuration.get(STREAMING_SOURCE_PARTITION_INCLUDE);
+		if (isStreamingSource()) {
+			Preconditions.checkArgument(
+					!configuration.contains(STREAMING_SOURCE_CONSUME_START_OFFSET),
+					String.format(
+							"The '%s' is not supported when set '%s' to 'latest'",
+							STREAMING_SOURCE_CONSUME_START_OFFSET.key(),
+							STREAMING_SOURCE_PARTITION_INCLUDE.key()));
+
+			Duration monitorInterval = configuration.get(STREAMING_SOURCE_MONITOR_INTERVAL) == null
+					? DEFAULT_LOOKUP_MONITOR_INTERVAL
+					: configuration.get(STREAMING_SOURCE_MONITOR_INTERVAL);
+			Preconditions.checkArgument(
+					monitorInterval.toMillis() >= DEFAULT_LOOKUP_MONITOR_INTERVAL.toMillis(),
+					String.format(
+							"Currently the value of '%s' is required bigger or equal to default value '%s' " +
+									"when set '%s' to 'latest', but actual is '%s'",
+							STREAMING_SOURCE_MONITOR_INTERVAL.key(),
+							DEFAULT_LOOKUP_MONITOR_INTERVAL.toMillis(),
+							STREAMING_SOURCE_PARTITION_INCLUDE.key(),
+							monitorInterval.toMillis())
+			);
+
+			hiveTableReloadInterval = monitorInterval;
+		} else {
+			Preconditions.checkArgument(
```

Comment:
Does this mean we require `STREAMING_SOURCE_ENABLE` to be set in order to load latest partition in temporal join? IMHO this limitation is not very friendly because many users still want to be able to use their hive tables in batch analysis. I think we should only require `STREAMING_SOURCE_ENABLE` if users want to consume data in a streaming fashion, i.e. periodically monitor files/partitions and fetch data incrementally.

Reply:
Yes, because we think using partition project is better to load specific partition in batch analysis, and from the semantic, a batch table does not have the latest partition but streaming hive table has. And for batch, user can still use `LOOKUP_JOIN_CACHE_TTL` to set the reload interval.

Reply:
Problem is, when `STREAMING_SOURCE_ENABLE` is set, the table can no longer be used in a batch job. And given that not many users are using the real "streaming read", it basically means such tables can only be used in temporal join. I'm fine if this is by design, but still feel it's unfriendly and unnecessary.

IMO, any table can have "latest partition" as long as the partitions are comparable in some way. Whether the data should be read in batch or streaming mode is an orthogonal concept.